common/djangoapps/cors_csrf/middleware.py

```diff
@@ -22,35 +47,76 @@
 
 from django.conf import settings
 from django.middleware.csrf import CsrfViewMiddleware
+from django.core.exceptions import MiddlewareNotUsed, ImproperlyConfigured
 
 log = logging.getLogger(__name__)
 
 
+def is_cross_domain_request_allowed(request):
+    """Check whether we should allow the cross-domain request.
+
+    We allow a cross-domain request only if:
+
+    1) The request has an http referer header.
+    2) The request is made securely and the referer has "https://" as the protocol.
+    3) The referer domain has been whitelisted.
+
+    Arguments:
+        request (HttpRequest)
+
+    Returns:
+        bool
+
+    """
+    referer = request.META.get('HTTP_REFERER')
+    if not referer:
+        log.debug("Request does not have a referer, so it is not a cross-domain request.")
+        return False
+    referer_parts = urlparse.urlparse(referer)
+
+    domain_is_whitelisted = (
+        getattr(settings, 'CORS_ORIGIN_ALLOW_ALL', False) or
+        referer_parts.hostname in getattr(settings, 'CORS_ORIGIN_WHITELIST', [])
+    )
+    if not domain_is_whitelisted:
+        log.debug(
+            (
+                "Domain '%s' is not on the cross domain whitelist.  "
+                "Add the domain to `CORS_ORIGIN_WHITELIST` or set "
+                "`CORS_ORIGIN_ALLOW_ALL` to True in the settings."
+            ), referer_parts.hostname
+        )
+        return False
+
+    request_is_secure = (
+        (request.is_secure() and referer_parts.scheme == 'https') or
+        getattr(settings, 'CORS_ALLOW_INSECURE', False)
+    )
+    if not request_is_secure:
+        log.debug(
+            "Request is not secure, so we cannot send the CSRF token. "
+            "For testing purposes, you can disable this check by setting "
+            "`CORS_ALLOW_INSECURE` to True in the settings"
+        )
+        return False
+
+    return True
+
+
 class CorsCSRFMiddleware(CsrfViewMiddleware):
     """
     Middleware for handling CSRF checks with CORS requests
     """
-    def is_enabled(self, request):
-        """
-        Override the `is_enabled()` method to allow cross-domain HTTPS requests
-        """
+    def __init__(self):
+        """Disable the middleware if the feature flag is disabled. """
         if not settings.FEATURES.get('ENABLE_CORS_HEADERS'):
-            return False
-
-        referer = request.META.get('HTTP_REFERER')
-        if not referer:
-            return False
-        referer_parts = urlparse.urlparse(referer)
-
-        if referer_parts.hostname not in getattr(settings, 'CORS_ORIGIN_WHITELIST', []):
-            return False
-        if not request.is_secure() or referer_parts.scheme != 'https':
-            return False
```

Comment:
Note that here, we returned False to let the standard CSRF middleware handle the HTTP requests, since he was already properly handling the referer. With the changes, it would still be the behavior, but before handing the request over to the standard middleware, it would log "Request does not have a referer, so it is not a cross-domain request." - which isn't necessarily true in the case of a HTTP request.